### ISSUE TITLE
Dashboard UI for Triggers

### ISSUE THREAD

**Issue by skaegi** [tektoncd/dashboard#509]
I think the trigger work solves a really critical problem -- mapping events to pipelineruns and resources. This issue is to look at what we might do to formally add it to the dashboard and also integrate it in with any pipelineruns and pipelineresources the trigger template creates in response to an event.

**Comment by kimholmes** [comment]
It might be helpful to define what at trigger is. Sounds like it's an "event." Can we list some examples of such events? What is a "trigger template?"

Based on this description, it sounds like we will need a page with a table containing said triggers. What descriptive columns will be needed on the table? Based on that, we can figure out what this page would link to within the dasbhoard.

Also, based on this description, it looks like a piece of this (triggers) might reside on the actual resource that was created as a result of said trigger. Whether it be PLR, or any other resource.
Looking forward to further conversation on this.

**Comment by ncskier** [comment]
The triggers here refer to the Triggers project: https://github.com/tektoncd/triggers. The work is still in-progress (and pre-alpha), so the Custom Resources (EventListener, TriggerBinding, and TriggerTemplate) are still changing as we develop them.

When we prioritize this work item and someone picks it up, then I will be happy to help create the specifications for how the Dashboard should deal with Triggers resources 👍 

Note: we might find it helpful to break this into separate issues for the EventListener, TriggerBinding, and TriggerTemplate

**Comment by kimholmes** [comment]
Need to begin design work on this by the end of September.

**Comment by ncskier** [comment]
Here's an overview of the Triggers resources at the moment (the project is still in development, so it is subject to change). Much of this information is re-iterated from the [Triggers README's](https://github.com/tektoncd/triggers).

## Overview
The Triggers project enables users to set up Tekton PipelineRuns/TaskRuns that will execute (start running) when an event is received. A common example use-case for Triggers is automatically creating of a `PipelineResource`, and `PipelineRun` after receiving a webhook from GitHub.

There are three custom resources in the Triggers project: `TriggerTemplate`, `TriggerBinding`, and `EventListener`. We would probably want a "list" page for each resource type, and maybe a "details" page as well if a user clicks on an item in the list; for example, a page listing all of the `EventListener` resources, and a "details" page when the user clicks on an `EventListener`. 

## tl;dr (this comment is long and throws a lot of information at you)
Focus on the __Important information to display in the details page__ sections for the important information to include in the details page for each resource.

## 1. TriggerTemplate
The `TriggerTemplate` specifies the resources a user would like to automatically create when an event is received.

Here's an example `TriggerTemplate` from the Triggers' [examples](https://github.com/tektoncd/triggers/blob/master/examples/triggertemplates/triggertemplate.yaml).
```yaml

apiVersion: tekton.dev/v1alpha1
kind: TriggerTemplate
metadata:
  name: pipeline-template
spec:
  params:
    - name: gitrevision
      description: The git revision
      default: master
    - name: gitrepositoryurl
      description: The git repository url
    - name: message
      description: The message to print
      default: This is the default message
  resourcetemplates:
    - apiVersion: tekton.dev/v1alpha1
      kind: PipelineResource
      metadata:
        name: git-source-$(uid)
        labels:
          triggertemplated: "true"
          generatedBy: "triggers-example"
      spec:
        type: git
        params:
        - name: revision
          value: $(params.gitrevision)
        - name: url
          value: $(params.gitrepositoryurl)
    - apiVersion: tekton.dev/v1alpha1
      kind: PipelineRun
      metadata:
        generateName: simple-pipeline-run
        labels:
          triggertemplated: "true"
          generatedBy: "triggers-example"
      spec:
        pipelineRef:
            name: simple-pipeline
        params:
        - name: message
          value: $(params.message)
        resources:
        - name: git-source
          resourceRef:
            name: git-source-$(uid)
```
In this example, the `TriggerTemplate` defines a `PipelineResource` and `PipelineRun` that will be created when an event is received. This `TriggerTemplate` also defines three parameters (`gitrevision`, `gitrepositoryurl`, and `message`) that are used in the `resourcetemplates`. 

### Important information to display in the details page:
- `params` (this is the list of inputs to the `TriggerTemplate`)
  - each parameter must have a `name`, and can optionally have a `default` and/or `description` value
- `resourcetemplates` (this is the list of resources that the TriggerTemplate will create)
  - it would probably be helpful to display at least the `kind` of each `resourcetemplate`

## 2. TriggerBinding
When an event is received, it will have a payload with important data. The `TriggerBinding` specifies what data from the event payload will be passed to a `TriggerTemplate`. The `TriggerTemplate` passes this data to the `TriggerTemplate` as parameters (`params`)

Here's an example `TriggerBinding` from the Triggers' [examples](https://github.com/tektoncd/triggers/blob/master/examples/triggerbindings/triggerbinding.yaml).
```yaml
apiVersion: tekton.dev/v1alpha1
kind: TriggerBinding
metadata:
  name: pipeline-binding
spec:
  params:
    - name: gitrevision
      value: $(event.head_commit.id)
    - name: gitrepositoryurl
      value: $(event.repository.url)
```

### Important information to display in the details page:
- `params`; each parameter has a `name` and `value`
  - These `params` are different from the `TriggerTemplate`'s `params`

## 3. EventListener
The `EventListener` will probably have the most information out of all the Triggers resources, it puts all of the resources together into a working bundle.

The `EventListener` creates a Pod that listens for events; it specifies what `TriggerTemplate` and `TriggerBinding` should be used when an event is received. It also specifies static `params` to pass to the `TriggerTemplate` and a `validate` task to validate the incoming event's authenticity.

The `EventListener` can have multiple `TriggerTemplate` + `TriggerBinding` bundles, and these are listed as the `triggers` property.

Here's an example `EventListener` from the Triggers' [examples](https://github.com/tektoncd/triggers/blob/master/examples/eventlisteners/eventlistener.yaml).
```yaml
apiVersion: tekton.dev/v1alpha1
kind: EventListener
metadata:
  name: listener
spec:
  serviceAccountName: tekton-triggers-example-sa
  triggers:
    - binding:
        name: pipeline-binding
      template:
        name: pipeline-template
      params:
        - name: message
          value: Hello from the Triggers EventListener!
```

### Important information to display in the details page:
- `triggers` (this is a list of `TriggerTemplate` + `TriggerBinding` pairs that will be triggered when an event is received)
  - `triggers.binding` (this is the `TriggerBinding` associated with the `EventListener`'s trigger)
    - it might be helpful if this is able to link to the associated `TriggerBinding` description page
  - `triggers.template` (this is the `TriggerTemplate` associated with the `EventListener`'s trigger)
    - it might be helpful if this is able to link to the associated `TriggerTemplate` description page
  - `triggers.params` (this is a list of static parameters that are passed to the `TriggerTemplate`)
    - each parameter has a `name` and `value` (these `params` are different from the `TriggerTemplate`'s `params`)
  - `triggers.validate` (this is a Task that will validate the received event's authenticity)
    - this is still being developed/documented; more information [here](https://github.com/tektoncd/triggers/blob/master/docs/validate-event.md)
- `serviceAccountName`
- The `EventListener` will soon have Status information which will be important to display as well, but that is still a work in progress: https://github.com/tektoncd/triggers/issues/54

**Comment by kimholmes** [comment]
First pass at triggers. 
**Questions:** 

- Do we need create and delete on the main list view pages?
- For the detail pages, what parts are editable, if any?
- Do the trigger components have labels?
- Should the "code" tab exist? If so, should it be called anything except "code?"

### TriggerTemplate
main:
![trigger_template_main](https://media.github.ibm.com/user/9341/files/c1f9af80-e6a1-11e9-9451-ef563800b037)

Should the columns in the Resource Templates table be switched? Kind, then name?
detail:
![trigger_template_detail](https://media.github.ibm.com/user/9341/files/e48bc880-e6a1-11e9-9107-e5e3d7511ca2)
code:
![trigger_template_code](https://media.github.ibm.com/user/9341/files/156bfd80-e6a2-11e9-95f1-95f9e21aeaf0)

### TriggerBinding
main:
![trigger_binding_main](https://media.github.ibm.com/user/9341/files/2b79be00-e6a2-11e9-9296-8b3b43c4962a)
detail:
![trigger_binding_detail](https://media.github.ibm.com/user/9341/files/43e9d880-e6a2-11e9-9889-d826a294b8e8)
code:
![trigger_binding_code](https://media.github.ibm.com/user/9341/files/4c421380-e6a2-11e9-94db-fc5c8099bbc9)

### EventListeners
main:
![eventlistener_main](https://media.github.ibm.com/user/9341/files/83b0c000-e6a2-11e9-9ac7-cefc7e6d376d)

How many multiple Bundles could there be? How many Validate Triggers could there be?
detail:
![eventlistener_detail](https://media.github.ibm.com/user/9341/files/8b706480-e6a2-11e9-92d3-c144e6d26cc1)
code:
![eventlistener_code](https://media.github.ibm.com/user/9341/files/94613600-e6a2-11e9-82f5-917fd587d368)

**Comment by kimholmes** [comment]
Updates to screens per last review with dev team:
- For all detail pages, I changed the "code" tab to say "YAML." I also removed the edit button as these will be uneditable. 
For **TriggersTemplate** I removed the word "value" from the table headers. Also, the PLR in the Resource Template table will not be inked.
![Trigger_template_detail2](https://user-images.githubusercontent.com/21316329/66777877-284bb280-ee98-11e9-96a9-2b66019e2c87.png)

For **EventListeners** I added a status and Service Account to the list view:
![EventListener_main2](https://user-images.githubusercontent.com/21316329/66778019-88425900-ee98-11e9-9859-282732001fb8.png)
On the Detail page, I moved the Params into the first table and added triggers.validate which consists of the task.ref:
![EventListener_detail2](https://user-images.githubusercontent.com/21316329/66778111-c3dd2300-ee98-11e9-88a3-58bdae81191a.png)

**Comment by ncskier** [comment]
Thanks for the updates @kimholmes! I think the updated TriggerTemplate page looks really good 👍 

For the EventListeners list page, I like the addition of a Status column, but I'm not sure if we'll be able to use a checkmark icon as the value. Right now, the status just indicates that a pod is running, it does not indicate whether the latest build has been successful, or something like that (which is what the checkmark makes me think of).

For the EventListener's details page, I think that the table is an improvement, but the structure is still not exactly accurate, because each trigger can have multiple parameters. Also, the structure just changed a little to this (so the Validate is now called Interceptor):
```
spec:
  serviceAccountName: tekton-triggers-example-sa
  triggers:
    - name: foo-trig
      interceptor:
        header:
        - name: Foo-Trig-Header1
          value: string-value
        - name: Foo-Trig-Header2
          value:
          - array-val1
          - array-val2
        objectRef:
          kind: Service
          name: gh-validate
          apiVersion: v1
          namespace: default
      binding:
        name: pipeline-binding
      template:
        name: pipeline-template
      params:
      - name: message
        value: Hello from the Triggers EventListener!
```
I think that it might be beneficial to have a separate table for each `trigger`, and this table will be the list of `params` with columns for `name` and `value`.

Thanks again for the updates!

**Comment by kimholmes** [comment]
A question for above about the possibility of listing a table for each Trigger on the EventListener detail page. How many Triggers might appear here? Need to figure out if it might be too many to fit into the UI, in which case we'll need to alter that design.

**Comment by a-roberts** [comment]
PR for the API added that the panelling can leverage created 🎉

**Comment by mii-w** [comment]
Trigger Template
![Screen Shot 2019-11-05 at 10 03 05 AM](https://user-images.githubusercontent.com/51759906/68222066-6c723480-ffb8-11e9-8385-2a9fdd87f399.png)
![Screen Shot 2019-11-05 at 10 03 11 AM](https://user-images.githubusercontent.com/51759906/68222076-6da36180-ffb8-11e9-9828-857076753a27.png)
![Screen Shot 2019-11-05 at 10 33 40 AM](https://user-images.githubusercontent.com/51759906/68222084-72681580-ffb8-11e9-9742-b3d240504f2f.png)
![Screen Shot 2019-11-05 at 10 33 45 AM](https://user-images.githubusercontent.com/51759906/68222092-7431d900-ffb8-11e9-9736-76ef0b886b47.png)
![Screen Shot 2019-11-05 at 10 43 53 AM](https://user-images.githubusercontent.com/51759906/68222632-4c8f4080-ffb9-11e9-8751-5790d46f98ba.png)

**Comment by a-roberts** [comment]
Quick update on this, only the details pages to go now (I've got a PR in, @carlos-logro is working on another and then one more is up for grabs!)

**Comment by AlanGreene** [comment]
Closing this, remaining design updates tracked in https://github.com/tektoncd/dashboard/issues/937
